cinnabar/wrangle.py

```diff
@@ -22,209 +23,137 @@ def read_csv(filepath: pathlib.Path) -> dict:
-            DG_B = self.graph.nodes[edge[1]]["exp_DG"]
-            edge[2]["exp_DDG"] = DG_B - DG_A
-            dDG_A = self.graph.nodes[edge[0]]["exp_dDG"]
-            dDG_B = self.graph.nodes[edge[1]]["exp_dDG"]
-            edge[2]["exp_dDDG"] = (dDG_A**2 + dDG_B**2) ** 0.5
-
-        self.n_ligands = self.graph.number_of_nodes()
-        self.n_edges = self.graph.number_of_edges()
-        self.degree = self.n_edges / self.n_ligands
-
-        # check the graph has minimal connectivity
-        self.check_weakly_connected()
-        if not self.weakly_connected:
-            print("Graph is not connected enough to compute absolute values")
+        # unpack data dictionary
+        fe = cls()
+        for r in data['Calculated']:
+            fe.add_measurement(r)
+        for r in data['Experimental'].values():
+            fe.add_measurement(r)
+
+        return fe
+
+    def add_measurement(self, measurement: Union[RelativeMeasurement, AbsoluteMeasurement]):
+        """Add new observation to FEMap, modifies in place
+
+        Any other attributes on the measurement are used as annotations
+
+        Raises
+        ------
+        ValueError : if bad type given
+        """
+        if isinstance(measurement, AbsoluteMeasurement):
+            # coerse to relative to simplify logic
+            meas_ = RelativeMeasurement(ligandA='NULL',
+                                        ligandB=measurement.ligand,
+                                        DDG=measurement.DG,
+                                        uncertainty=measurement.uncertainty,
+                                        computational=measurement.computational)
+        elif isinstance(measurement, RelativeMeasurement):
+            meas_ = measurement
         else:
-            self.generate_absolute_values()
+            raise TypeError()
+
+        # slurp out tasty data, anything but labels
+        d = dict(meas_)
+        d.pop('ligandA', None)
+        d.pop('ligandB', None)
+        d.pop('ligand', None)
 
-    def check_weakly_connected(self):
-        undirected_graph = self.graph.to_undirected()
-        self.weakly_connected = nx.is_connected(undirected_graph)
+        if meas_.computational:
+            self.computational_graph.add_edge(meas_.ligandA, meas_.ligandB, **d)
+        else:
+            self.experimental_graph.add_edge(meas_.ligandA, meas_.ligandB, **d)
+
+    @property
+    def n_measurements(self) -> int:
+        """Total number of both experimental and computational measurements"""
+        return len(self.experimental_graph.edges) + len(self.computational_graph.edges)
+
+    @property
+    def n_ligands(self) -> int:
+        """Total number of unique ligands"""
+        return len(self.experimental_graph.nodes | self.computational_graph.nodes)
+
+    @property
+    def degree(self) -> float:
+        """Average degree of all nodes"""
+        return self.n_measurements / self.n_ligands
+
+    def is_weakly_connected(self) -> bool:
```

Comment:
long term we should consider calling this just "is_connected" given that's exactly what does, you can't really tell if it's weakly or strongly connected?